File: ios/Packages/FinancialDomain/Sources/FinancialDomain/Models/Reports.swift
import Foundation

public struct ReportCategory: Sendable, Hashable, Identifiable {
    public var id: String { name }
    public var name: String
    public var amount: Money

    public init(name: String, amount: Money) {
        self.name = name
        self.amount = amount
    }
}

public struct ReportAccountRef: Sendable, Hashable, Identifiable {
    public var id: String
    public var name: String
    public var type: String

    public init(id: String, name: String, type: String = "BANK") {
        self.id = id
        self.name = name
        self.type = type
    }
}

public struct ReportsSnapshot: Sendable, Hashable {
    public var months: Int
    public var selectedMonth: YearMonth
    public var income: Money
    public var expense: Money
    public var categories: [ReportCategory]
    public var accounts: [ReportAccountRef]
    public var calculationVersion: String?

    public init(
        months: Int,
        selectedMonth: YearMonth,
        income: Money,
        expense: Money,
        categories: [ReportCategory],
        accounts: [ReportAccountRef],
        calculationVersion: String? = nil
    ) {
        self.months = months
        self.selectedMonth = selectedMonth
        self.income = income
        self.expense = expense
        self.categories = categories
        self.accounts = accounts
        self.calculationVersion = calculationVersion
    }
}

File: supabase/functions/pluggy-proxy/handlers/reports.ts
/**
 * Reports aggregations — web analytics.js (exclude bill payments).
 */
import { errorResponse, jsonResponse } from "../middleware/http.ts";
import { type PluggyClient } from "./pluggy.ts";
import {
  buildIncomeExpenseSeries,
  currentYm,
  expensesByCategory,
  isExpenseTx,
  isIncomeTx,
  lastNMonths,
  monthCashflow,
} from "../utils/dashboardAnalytics.ts";
import {
  fetchAccountsWithConnectors,
  fetchAllTransactionsForAccount,
  loadManualTransactions,
} from "../utils/creditLedger.ts";

export async function handleReports(client: PluggyClient, url: URL): Promise<Response> {
  const months = Math.min(12, Math.max(3, Number(url.searchParams.get("months") || 6) || 6));
  const accountId = url.searchParams.get("accountId") || "";
  const ym = currentYm();

  const [pluggyAccounts, manuals] = await Promise.all([
    fetchAccountsWithConnectors(client),
    loadManualTransactions(client),
  ]);
  const targets = accountId
    ? pluggyAccounts.filter((a) => String(a.id) === accountId)
    : pluggyAccounts;
  const pluggyTxs = (await Promise.all(
    targets.map((a) => fetchAllTransactionsForAccount(client, String(a.id))),
  )).flat();
  const fromYm = lastNMonths(months, ym)[0];
  const transactions = [...pluggyTxs, ...manuals].filter((t) => {
    if (accountId && String(t.accountId) !== accountId && !t.isManual) return false;
    const tYm = String(t.date || "").slice(0, 7);
    return tYm >= fromYm && tYm <= ym;
  });

  let income = 0;
  let expense = 0;
  transactions.forEach((t) => {
    if (isIncomeTx(t)) income += Math.abs(Number(t.amount) || 0);
    else if (isExpenseTx(t)) expense += Math.abs(Number(t.amount) || 0);
  });
  const cashflow = monthCashflow(transactions, ym);
  const incomeExpenseSeries = buildIncomeExpenseSeries(transactions, months, ym);
  const categories = expensesByCategory(transactions, { limit: 20 });

  return jsonResponse({
    months,
    selectedMonth: ym,
    income: Number(income.toFixed(2)),
    expense: Number(expense.toFixed(2)),
    cashflow,
    incomeExpenseSeries,
    categories,
    accounts: pluggyAccounts.map((a) => ({
      id: String(a.id),
      name: String(a.marketingName || a.name || "Conta"),
      type: String(a.type || "BANK"),
    })),
  });
}
